src/OpenGames/Preprocessor/THSyntax.hs

```diff
@@ -0,0 +1,131 @@
+        l2 = Reindex (FlattenTuples (length lines)) (foldl1 Sequential (map compileLine lines))
+        l3 = Lens (Lambda covariantBlockContext (Expressions (map (VarE . mkName) $ blockCovariantOutputs block)))
+                  (Curry (Multiplex covariantBlockContext (Variables (blockContravariantInputs block))))
+
+
+
+newtype AtomExpression = AtomExpression Exp
+
+flattenVariables = Variables . concat . map vars
+
+sequentialTrivialL, sequentialTrivialR, simultaneousTrivialL, simultaneousTrivialR :: FreeOpenGame -> FreeOpenGame -> FreeOpenGame
+sequentialTrivialL g h   = Reindex UnitIntroL (Sequential g h)
+sequentialTrivialR g h   = Reindex UnitIntroR (Sequential g h)
+simultaneousTrivialL g h = Reindex UnitIntroL (Simultaneous g h)
+simultaneousTrivialR g h = Reindex UnitIntroR (Simultaneous g h)
+
+covariantContexts :: Block -> [Variables]
+covariantContexts block = map f (init (inits (map (Variables . covariantOutputs) (blockLines block))))
+  where f contexts = flattenVariables (Variables (blockCovariantInputs block) : contexts)
+
+
+contravariantContexts :: Block -> [Variables]
+contravariantContexts block = map (f . reverse) (tail (tails (map (Variables . contravariantOutputs) (blockLines block))))
+  where f contexts = flattenVariables (concat [[Variables (blockCovariantInputs block)],
+                                                   map (Variables . covariantOutputs) (blockLines block),
+                                                   [Variables (blockContravariantInputs block)],
+                                                   contexts])
+
+linesWithContext :: Block -> [LineWithContext]
+linesWithContext block = zipWith3 LineWithContext (blockLines block) (covariantContexts block) (contravariantContexts block)
+
+param :: String -> Q Exp
+param = pure . VarE . mkName
+
+compileQLine :: QLine -> Q SLine
+compileQLine qline = do covIn <- traverse id $ qcovariantInputs qline
+                        conIn <- traverse id $ qcontravariantInputs qline
+                        exp <- qmatrix qline
+                        let covOut = qcovariantOutputs qline
+                        let conOut = qcontravariantOutputs qline
+                        pure $ SLine covIn conOut exp covOut conIn
+
+
+toBlock :: ToLine l => GBlock l -> Q Block
+toBlock (GBlock a b lines c d) = do l' <- traverse toLine lines
+                                    pure $ GBlock a b l' c d
+
+generateGame' :: ToLine line => String -> [String] -> GBlock line -> Q [Dec]
+generateGame' name args block =
+  do blk <- toBlock block
+     game <- interpretOpenGame (compileBlock blk)
+     pure $ [FunD (mkName name) [Clause (fmap (VarP . mkName) args) (NormalB game) []]]
+
+generateGame :: String -> [QLine] -> Q [Dec]
+generateGame name lines = do lines <- traverse compileQLine lines
+                             generateGame' name [] $ GBlock [] [] lines [] []
```

Comment:
Those could probably be typeclasses instanciated with `GBlock line` and `[QLine]`